### PR TITLE
task(content): Add support for tracing on client side

### PR DESCRIPTION
**Pull request by dschom**
## Because

- We want to start trace capture at client side

## This pull request

- Adds support for client side tracing in tracing/browser-tracing.ts
- Some refactors to fxa-shared/tracing to ensure code is reused
- Cleans up tests
- Introduces traceparent header and tracestate html headers
- Opens up OTLP ports on jaeger
- Introduces a clientName to trace config

## Issue that this pull request solves

Closes: FXA-5085

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

image

_Note that above screen shot starts with a client side request._

## Other information (Optional)

To do exploratory testing locally, add the following values to a .env file in the root folder, and run `yarn stop && dotenv -- yarn start`. Then start using fxa and navigate to http://localhost:16686 to explore traces.

TRACING_SAMPLE_RATE=1
TRACING_BATCH_PROCESSING=false
TRACING_JAEGER_EXPORTER_ENABLED=true
TRACING_CONSOLE_EXPORTER_ENABLED=true

### CODE REVIEW

File: packages/fxa-content-server/app/scripts/app.js [hunk 1]
```diff
@@ -1,12 +1,18 @@
...
-import 'core-js'
+import 'core-js';
+import * as tracing from 'fxa-shared/tracing/browser-tracing';
 import ConfigLoader from './lib/config-loader';
 import AppStart from './lib/app-start';
 
 const configLoader = new ConfigLoader();
 configLoader.fetch().then((config) => {
+  if (config.tracing) {
+    const flowId = $ && $(document.body).data('flowId'); // eslint-disable-line no-undef
```
Comment (chenba) [inline packages/fxa-content-server/app/scripts/app.js:12]: Is this the flowId from the flow metrics?

Comment (dschom) [reply packages/fxa-content-server/app/scripts/app.js:12]: Yes. We wanted to add this so we could aggregate traces across a flow. If there is a better way to get this value, I'd love to know.

File: packages/fxa-content-server/server/lib/beta-settings.js [hunk 2]
```diff
@@ -75,6 +77,15 @@ function swapBetaMeta(html, metaContent = {}) {
...
   });
 
+  Object.keys(metaContent).forEach((key) => {
+    let val = metaContent[key];
+    if (typeof val === 'object') {
+      val = encodeURIComponent(JSON.stringify(metaContent[key]));
+    }
+
+    result = result.replace(key, val);
+  });
```
Comment (chenba) [inline packages/fxa-content-server/server/lib/beta-settings.js:87]: nit: I think we can combine this with the `forEach` above.

File: packages/fxa-content-server/server/lib/csp/blocking.js [hunk 3]
```diff
@@ -37,6 +37,13 @@ module.exports = function (config) {
   const GOOGLE_AUTH = 'https://accounts.google.com';
   const APPLE_AUTH = 'https://appleid.apple.com';
 
+  const tracingConfig = config.get('tracing');
+  let TRACE_OTLP_URL = undefined;
+  if (tracingConfig.jaeger && tracingConfig.jaeger.otlpUrl) {
+    TRACE_OTLP_URL = url.parse(tracingConfig.jaeger.otlpUrl);
+    TRACE_OTLP_URL = TRACE_OTLP_URL.href.replace(TRACE_OTLP_URL.pathname, '');
```
Comment (chenba) [inline packages/fxa-content-server/server/lib/csp/blocking.js:44]: What is this about?

Comment (dschom) [reply packages/fxa-content-server/server/lib/csp/blocking.js:44]: This allows the cors request through. We already have a a setting which indicates the full URL of the tracing endpoint, so this just extracting the root of the URL of that end point to configure CORS settings. 

Maybe it'd be more clear to just have a separate config option for this?

**Comment by chenba** [comment]
Empty file: https://github.com/mozilla/fxa/pull/14212/files#diff-268ccd6b5a71c3f7fb158ef2525acf49b7518c53a16e0efde38efd1b0817d22f ?

File: packages/fxa-content-server/server/lib/csp/blocking.js (hunk 3 above)
Comment (chenba) [reply packages/fxa-content-server/server/lib/csp/blocking.js:44]: Oh, nah.  It was my bad; I commented before reading farther to understand how that was used.

**Review by chenba — APPROVED** [review]
Obviously I don't know the OpenTelemetry APIs enough to really review this, but looks good.  🏆 on the tests.  Thanks, Dan.

**Pull request merged**